magicrescue.1
.\" Automatically generated by Pod::Man v1.37, Pod::Parser v1.14
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used to
.\" do unbreakable dashes and therefore won't be available.  \*(C` and \*(C'
.\" expand to `' in nroff, nothing in troff, for use with C<>.
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "MAGICRESCUE 1"
.TH MAGICRESCUE 1 "2004-04-30" "1.1.1" "Magic Rescue"
.SH "NAME"
magicrescue \- Scans a block device and extracts known file types by looking at
magic bytes.
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBmagicrescue\fR [ \fIoptions\fR ] \fIdevices\fR
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
Magic Rescue opens \fIdevices\fR for reading, scans them for file types it knows
how to recover and calls an external program to extract them.  It looks at
\&\*(L"magic bytes\*(R" in file contents, so it can be used both as an undelete utility
and for recovering a corrupted drive or partition.  It works on any file system,
but on very fragmented file systems it can only recover the first chunk of
each file.  These chunks are sometimes as big as 50MB, however.
.PP
To invoke \fBmagicrescue\fR, you must specify at least one device and the \fB\-d\fR
and \fB\-r\fR options.  See the \*(L"\s-1USAGE\s0\*(R" section in this manual for getting
started.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-b\fR \fIblocksize\fR" 7
.IX Item "-b blocksize"
Default: 1.  This will direct \fBmagicrescue\fR to only consider files that start
at a multiple of the \fIblocksize\fR argument.  The option applies only to the
recipes following it, so by specifying it multiple times it can be used to get
different behavior for different recipes.
.Sp
Using this option you can usually get better performance, but fewer files will
be found.  In particular, files with leading garbage (e.g. many mp3 files) and
files contained inside other files are likely to be skipped.  Also, some file
systems don't align small files to block boundaries, so those won't be found
this way either.
.Sp
If you don't know your file system's block size, just use the value 512, which
is almost always the hardware sector size.
.IP "\fB\-d\fR \fIdirectory\fR" 7
.IX Item "-d directory"
Mandatory.  Output directory for found files.  Make sure you have plenty of free
space in this directory, especially when extracting very common file types such
as jpeg or gzip files.  Also make sure the file system is able to handle
thousands of files in a single directory, i.e. don't use \s-1FAT\s0 if you are
extracting many files.
.Sp
You should not place the output directory on the same block device you are
trying to rescue files from.  This might add the same file to the block device
ahead of the current reading position, causing \fBmagicrescue\fR to find the same
file again later.  In the worst theoretical case, this could cause an infinite
loop where the same file is extracted thousands of times until disk space is
exhausted.  You are also likely to overwrite the deleted files you were looking
for in the first place.
.IP "\fB\-r\fR \fIrecipe\fR" 7
.IX Item "-r recipe"
Mandatory.  Recipe name, file, or directory.  Specify this as either a plain
name (e.g.  \f(CW\*(C`jpeg\-jfif\*(C'\fR) or a path (e.g. \fIrecipes/jpeg\-jfif\fR).  If it doesn't
find such a file in the current directory, it will look in \fI./recipes\fR and
\&\fIPREFIX/share/magicrescue/recipes\fR, where \fI\s-1PREFIX\s0\fR is the path you installed
to, e.g. \fI/usr/local\fR.
.Sp
If \fIrecipe\fR is a directory, all files in that directory will be treated as
recipes.
.Sp
Browse the \fIPREFIX/share/magicrescue/recipes\fR directory to see what recipes
are available.  A recipe is a text file, and you should read the comments
inside it before using it.  Either use the recipe as it is or copy it somewhere
and modify it.
.Sp
For information on creating your own recipes, see the \*(L"\s-1RECIPES\s0\*(R" section.
.IP "\fB\-M\fR \fIoutput_mode\fR" 7
.IX Item "-M output_mode"
Produce machine-readable output to stdout.  \fIoutput_mode\fR can be:
.RS 7
.IP "\fBi\fR" 4
.IX Item "i"
Print each input file name before processing
.IP "\fBo\fR" 4
.IX Item "o"
Print each output file name after processing
.IP "\fBio\fR" 4
.IX Item "io"
Print both input and output file names.  Input file names will be prefixed by
\&\f(CW\*(C`i\*(C'\fR and a space.  Output file names will be prefixed by \f(CW\*(C`o\*(C'\fR and a space.
.RE
.RS 7
.Sp
Nothing else will be written to standard output in this mode.
.RE
.SH "USAGE"
.IX Header "USAGE"
Say you have destroyed the file system on /dev/hdb1 and you want to extract
all the jpeg files you lost.  This guide assumes you have installed Magic
Rescue in \fI/usr/local\fR, which is the default.
.PP
Make sure \s-1DMA\s0 and other optimizations are enabled on your disk, or it will take
hours.  In Linux, use hdparm to set these options:
.PP
.Vb 1
\&    $ hdparm -d 1 -c 1 -u 1 /dev/hdb
.Ve
.PP
Choose your output directory, somewhere with lots of disk space.
.PP
.Vb 1
\&    $ mkdir ~/output
.Ve
.PP
Look in the \fI/usr/local/share/magicrescue/recipes\fR directory for the recipes
you want.  Magic Rescue somes with recipes for some common file types, and you
can make your own too (see the next section).  Open the recipes you want to use
in a text editor and read their comments.  Most recipes require 3rd party
software to work, and you may want to modify some parameters (such as
\&\fBmin_output_file\fR) to suit your needs.
.PP
Then invoke \fBmagicrescue\fR
.PP
.Vb 1
\&    $ magicrescue -r jpeg-jfif -r jpeg-exif -d ~/output /dev/hdb1
.Ve
.PP
It will scan through your entire hard disk, so it may take a while.  You can
stop it and resume later of you want to.  Interrupt it (with \s-1CTRL+C\s0) and note
the progress information saying what address it got to.  Then restart it later
with:
.PP
.Vb 2
\&    $ /usr/local/share/magicrescue/tools/inputseek 0x2F3A94BE0 \e
\&      magicrescue -r jpeg-jfif -r jpeg-exif -d ~/output - < /dev/hdb1
.Ve
.PP
When it has finished you will probably find thousands of .jpg files in
\&\fI~/output\fR, including things you never knew was in your browser cache.  Sorting
through all those files can be a huge task, so you may want to use software or
scripts to do it.
.PP
First, try to eliminate duplicates with the \fBdupemap\fR(1) tool included in this
package.
.PP
.Vb 1
\&    $ dupemap delete,report ~/output
.Ve
.PP
If you are performing an undelete operation you will want to get rid
of all the rescued files that also appear on the live file system.  See the
\&\fBdupemap\fR(1) manual for instructions on doing this.
.PP
If that's not enough, you can use use \fBmagicsort\fR to get a better overview:
.PP
.Vb 1
\&    $ magicsort ~/output
.Ve
.SH "RECIPES"
.IX Header "RECIPES"
.Sh "Creating recipe files"
.IX Subsection "Creating recipe files"

NEWS
Version 1.1.9: Mon Apr 12 20:21:16 CEST 2010
- Recipe for rar archives

Version 1.1.8: Wed Nov  4 16:12:53 CET 2009
- Add mbox recipes, contributed by Jeff Bucove.
- Raise maximum virtual memory usage for helper programs from 50MB to 1024MB.
  It turns out that mencoder needs around 150MB on 64-bit systems.
- Place tools directory first in path to avoid name clashes with other
  programs.

Version 1.1.7: Fri Sep  4 19:48:14 CEST 2009
- Update zip recipe to work with latest version of Zip and latest
  OpenOffice.org file formats (thanks to Fernando)

Version 1.1.6: Sun Feb 15 22:13:54 CET 2009
- Recipe for Nikon raw photos
- Recipe for large PPM files

Version 1.1.5: Thu Jan 17 14:03:49 CET 2008
- Recipe for Flac files
- Recipe for Canon CR2 RAW files
- Configure script compatibility fix for Debian-based Linuxes

Version 1.1.4: Sun Aug 15 01:11:27 CEST 2004
- Fixed a minor bug that prevented some not yet written recipes from working
  perfectly

Version 1.1.3: Tue Aug 10 22:32:42 CEST 2004
- Fixed a minor bug to make jpeg matching 100% reliable instead of 99.4%
- Added GIMP xcf recipe
- Added text extraction and experimental recipes for perl and gpl
- Added test suite to ensure correctness
- Support for dupemap on Debian and others
- Extraction of msoffice on cygwin

Version 1.1.2: Thu May  6 22:47:52 CEST 2004
- Added the -O option for easy resuming from any offset
- Added the -I option for easier usage with find(1)
- Reorganized source code
- Wrote more documentation

Version 1.1.1: Sat May  1 19:14:14 CEST 2004
- Added recipe for zip files and OpenOffice.org documents
- Added -b option for block alignment
- If argument to -r is a directory, read all files within it
- Anti-overlapping between files is now global instead of per-recipe
- Fixed bug that made anti-overlap code skip too much sometimes
- Fixed bug that prevented the use of an absolute path for recipes
- dupemap works on KNOPPIX now

Version 1.1.0: Tue Apr 13 13:23:28 CEST 2004
- Transformed README files into proper manpages and wrote more documentation
- Added dupemap, another tool for getting a better overview of the output
- Recipes take the output file name in $1 now, instead of $0
- Changed the output file format to adapt to "find|xargs magicrescue" usage
- Make magicrescue aware of output-file renaming
- Ported to Solaris
- Added more choices for machine-readable output

Version 1.0.4: Fri Mar 26 18:17:05 CET 2004
- tools/checkrecipe was broken for some file types, fixed that
- Fixed a rounding bug that could lead to segfaulting on some rare data
- Limits memory to 50MB per process, to prevent helper programs from exhausting
  memory on invalid input

Version 1.0.3: Tue Mar 23 16:02:19 CET 2004
- New recipes for Microsoft Office formats and ELF
- Recipe for gzip now extracts file names, if available
- Proper support was added for resuming an interrupted run
- Output file names now reflect match offsets
- Portability fixes, especially for Cygwin

Version 1.0.2: Mon Mar 15 13:17:48 CET 2004
- Much improved mp3 recipe
- Anti-overlapping prevents the same recipe from extracting part of the same
  file twice
- Added magicsort for categorizing the output files
- Magic Rescue now has a "normal" build system (./configure, make, make install)
- Bugfixes in the buffer overlapping code
- Prevents helper programs from leaving core files

Version 1.0.1: Fri Mar  5 11:15:04 CET 2004
- Nicer memory management
- Improved .avi recipe
- mp3 recipe
- gzip recipe

Version 1.0.0: Mon Mar  1 21:56:45 CET 2004
- Initial public release
- Has recipe parser and comes with recipes for avi, jpeg and png



src/magicrescue.c
/*
 * Magic Rescue main program
 * Copyright (C) 2004 Jonas Jensen <jbj@knef.dk>
 * 
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */
#include "config.h"

#ifdef HAVE_GETRLIMIT
# include <sys/time.h>
# include <sys/resource.h>
#endif /* HAVE_GETRLIMIT */

#include <sys/stat.h>
#include <sys/types.h>

#include <dirent.h>
#include <errno.h>
#include <fcntl.h>
#include <limits.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "util.h"
#include "array.h"
#include "recipe.h"
#include "scanner.h"
#include "magicrescue.h"

struct range {
    off_t begin;
    off_t end;
};

/* An array of struct scanner.  Each scanner can contain one or more recipes. */
static struct array scanners;

/* An array of struct range.  Used by the anti-overlap code to list ranges that
 * have already been extracted and should not be extracted again.
 * A value of begin < 0 means that a slot is unused. */
static struct array skip_ranges;

char *output_dir = NULL;
enum OUTPUT_MODE machine_output = OUT_HUMAN;
enum NAME_MODE name_mode = MODE_DEVICE;

/* Structure containing the current input file name and position */
struct progress progress;

/* This value is the max length of a chunk of bytes to be looked at when
 * scanning.  It is constant after all recipes have been parsed. */
static ssize_t overlap = 0;

/* General-purpose buffer.  Mostly used to hold the raw data when scanning. */
static char *buf;
static const size_t bufsize = 102400;


/* Signal handler for fatal signals */
static void signal_beforedeath(int signo)
{
    fprintf(stderr,
"\nmagicrescue: killed by signal %d at offset 0x%llX in file %s\n",
	    signo, (long long)progress.position, progress.device);

    if (progress.position)
	fprintf(stderr,
"Use the -O option to resume from this offset later\n");

    exit(0);
}

/* Add a scanner to the global scanners list, or reuses an existing entry if
 * an identical one exists.
 * Returns a pointer to the added scanner. */
static struct scanner *scanner_add_or_reuse(struct scanner *scanner)
{
    struct scanner *el;

    array_foreach(&scanners, el) {
	if (scanner_compare(scanner, el) == 0) {
	    scanner_destroy(scanner);
	    return el;
	}
    }

    return array_add(&scanners, scanner);
}

/* Free the global scanners list. */
static void scanners_free(void)
{
    struct scanner *scanner;

    array_foreach(&scanners, scanner) {
	scanner_destroy(scanner);
    }
    array_destroy(&scanners);
}

/* Scans scanbuf with all scanners in the global list, testing all recipes on
 * them.
 * Returns 1 if an external program has been called, which means the file
 * position has changed.  Returns -1 on fatal error. */
static int scan_buf(const char *scanbuf, ssize_t scanbuf_len,
	int fd, off_t scanbuf_filepos)
{
    int filepos_changed = 0;
    struct scanner *scanner;
    struct recipe *r;
    struct range *range;

    /* clean up ranges we have gone past, and see if we can skip some of the
     * buffer */
    array_foreach(&skip_ranges, range) {
	if (range->begin < 0) continue;

	if (range->end < scanbuf_filepos) {
	    range->begin = -1;

	} else if (range->end > scanbuf_filepos + scanbuf_len) {
	    return 0;

	} else if (range->begin <= scanbuf_filepos
		&& range->end > scanbuf_filepos) {
	    scanbuf         += range->end - scanbuf_filepos;
	    scanbuf_len     -= range->end - scanbuf_filepos;
	    scanbuf_filepos  = range->end;
	}
    }

    array_foreach(&scanners, scanner) {
	const char *p = scanbuf + scanner->offset;

	while (p - scanbuf < scanbuf_len &&
		(p = scanner->func(p, scanbuf_len - (p-scanbuf) +
				   scanner->offset,
				   &scanner->param,
				   scanbuf_filepos + (p-scanbuf)))) {
	    off_t filepos;

	    p -= scanner->offset;
	    filepos = scanbuf_filepos + (p-scanbuf);

	    array_foreach(&skip_ranges, range) {
		if (range->begin >= 0
			&& filepos > range->begin
			&& filepos < range->end) {
		    goto keep_scanning;
		}
	    }

	    array_foreach(&scanner->recipes, r) {
		struct operation *op;
		int ops_matched = 1;

		array_foreach(&r->ops, op) {
		    if (!op->func(p + op->offset, &op->param)) {
			ops_matched = 0;
			break;
		    }
		}

		if (ops_matched) {
		    off_t output_size;

		    fprintf(stderr, "Found %s at 0x%llX\n",
			    r->name, (long long)filepos);

		    output_size = extract(fd, r, filepos);
		    filepos_changed = 1;

		    if (output_size < 0)
			return -1;

		    if (r->allow_overlap >= 0
			    && output_size > r->allow_overlap) {

			struct range *tmp;
			range = NULL;
			
			/* find an available range, or add one */
			array_foreach(&skip_ranges, tmp) {
			    if (tmp->begin < 0) {
				range = tmp;
			    }
			}
			if (!range)
			    range = array_add(&skip_ranges, NULL);

			range->begin = filepos;
			range->end   = filepos + output_size - r->allow_overlap;

			break; /* try no more recipes */
		    }
		}
	    }

keep_scanning:
	    p += scanner->offset + 1;
	}
    }

    return filepos_changed;
}

/* Opens and scans an entire device for matches with the selected scanners. */
static void scan_disk(const char *device, const char *offset)
{
    ssize_t got;
    int fd = 0;
    int result, firsttime = 1;

    char *readbuf = buf, *scanbuf = buf;
    size_t readsize = bufsize;
    off_t offset_before_read;
    
    if (strcmp(device, "-") != 0 &&
	    (fd = open(device, O_RDONLY)) == -1) {
	fprintf(stderr, "opening %s: %s\n", device, strerror(errno));
	return;
    }

    offset_before_read = rich_seek(fd, offset);
    if (offset_before_read == (off_t)-1) {
	fprintf(stderr, "lseek failed on %s: %s\n", device, strerror(errno));
	close(fd);
	return;
    }

    snprintf(progress.device, sizeof progress.device, "%s", device);
    {
	char *tmp = strrchr(device, '/');
	snprintf(progress.device_basename, sizeof progress.device_basename, 
		"%s", tmp ? tmp+1 : device);
    }

    if ((machine_output & OUT_IO) == OUT_IO)
	printf("i %s\n", device);
    else if (machine_output & OUT_I)
	printf("%s\n", device);

    /* In the first iteration, the scan buffer and read buffer overlap like
     * this:
     * SSSSSSSSSSSSSSSSS
     * RRRRRRRRRRRRRRRRRRRRR
     *
     * For the rest of the iterations they overlap like this:
     * SSSSSSSSSSSSSSSSS
     *     RRRRRRRRRRRRRRRRR
     *
     * After each iteration the unscanned part of readbuf will be copied to the
     * start of scanbuf, and new data will be read into readbuf */

    while ((got = read(fd, readbuf, readsize)) > overlap) {
	progress.position = offset_before_read;

	result = scan_buf(scanbuf, got + (readbuf - scanbuf) - overlap,
		fd, offset_before_read - (readbuf - scanbuf));
	if (result == -1) {
	    close(fd);
	    return;

	} else if (result) {
	    if (lseek(fd, offset_before_read + got, SEEK_SET) == (off_t)-1) {
		perror("lseek()");
		close(fd);
		return;
	    }
	}

	memcpy(scanbuf, readbuf + got - overlap, overlap);

	if (firsttime) {
	    firsttime = 0;
	    readbuf  += overlap;
	    readsize -= overlap;
	}

	offset_before_read += got;
    }

    if (got == -1) {
	fprintf(stderr, "Read error on %s at %lld bytes: %s\n",
		device, (long long int)offset_before_read, strerror(errno));

	if (errno == EIO) {
	    fprintf(stderr, "Note that on some block devices this just "
		    "means end of file.\n");
	}
    } else {
	fprintf(stderr, "Scanning %s finished at %lldMB\n",
		device, (long long int)(offset_before_read >> 20));
    }

    close(fd);

    /* Reset all ranges after processing a file */
    {
	struct range *range;
	array_foreach(&skip_ranges, range) {
	    range->begin = -1;
	}
    }
}

/* Parse a recipe file, adding it to the global list of scanners.  If
 * global_scanner is not NULL, it points to a scanner that will be forced into
 * this recipe.
 * Returns 0 on success, -1 on fatal error. */
static int parse_recipe(const char *recipefile, 
	struct scanner *global_scanner)
{
    struct scanner *scanner = global_scanner;
    struct recipe r;
    FILE *fh = NULL; 
    char **sp, *search_path[] = { "recipes", NULL, NULL };

#ifdef RECIPE_PATH
    search_path[1] = RECIPE_PATH;
#endif

    fh = fopen(recipefile, "r");
    for (sp = search_path; !fh && *sp; sp++) {
	char path[PATH_MAX];
	snprintf(path, sizeof path, "%s/%s", *sp, recipefile);
	fh = fopen(path, "r");
    }
    
    if (!fh) {
	fprintf(stderr, "Opening %s: %s\n", recipefile, strerror(errno));
	return -1;
    }

    recipe_init(&r);

    {
	const char *basename = strrchr(recipefile, '/');
	r.name = strdup(basename ? basename + 1 : recipefile);
    }

    while (fgets(buf, bufsize, fh)) {
	char opname[64];
	int magic_offset, param_offset;
	buf[strlen(buf) - 1] = '\0'; /* kill trailing newline */

	if (sscanf(buf, "%d %63s %n",
		    &magic_offset, opname, &param_offset) >= 2) {
	    ssize_t len = 0;
	    char *param = buf + param_offset;
	    
	    if (!scanner) {
		/* Try to make it the scanner function */
		
		if (strcmp(opname, "string") == 0
			|| strcmp(opname, "char") == 0) {
		    struct string string;
		    string_init(&string, buf + param_offset);

		    scanner = scanner_new();
		    scanner->offset = magic_offset;

		    if (string.l == 1) {
			scanner->func = scanner_char;
			scanner->param.c = string.s[0];
			string_destroy(&string);
			len = 1;
		    } else if (string.l > 1) {
			scanner->func = scanner_string;
			scanner->param.scanstring.string = string;
			scanner_string_init(&scanner->param);
			len = string.l;
		    }
		}

		if (scanner) {
		    scanner->extra_len = scanner->offset + len - 1;
		}
	    }

	    if (len <= 0) {
		/* If it did not become a scanner function it will become an
		 * operation */
		struct operation *op = array_add(&r.ops, NULL);

		op->offset = magic_offset;

		if (strcmp(opname, "string") == 0
			|| strcmp(opname, "char") == 0) {
		    op->func = op_string;
		    string_init(&op->param.string, param);
		    len = op->param.string.l;

		} else if (strcmp(opname, "int32") == 0) {
		    char buf_val[16], buf_mask[16];
		    op->func = op_int32;
		    if (sscanf(param, "%15s %15s", buf_val, buf_mask) != 2) {
			fprintf(stderr, "%s: invalid parameter for %s: %s\n",
				recipefile, opname, param);
			fclose(fh);
			return -1;
		    }
		    op->param.int32.val  = hextoll(buf_val);
		    op->param.int32.mask = hextoll(buf_mask);
		    op->param.int32.val &= op->param.int32.mask;
		    len = 4;

		} /*
		     To add more operations:
		     - Add another type to union param if you have to.
		     - Define a test function like op_string that returns 1 on
		       success and 0 on failure.
		     - Add another "else if" clause here.  It must assign to
		       op->func and to len.
		     - The value assigned to len should be the maximum number of
		       bytes your test function might read.  Never read more
		       bytes than this.
		     - Add a case to op_destroy of cleanup is needed
		     The test functions you add should be general and reusable,
		     such as string, int, md5, regexp.  To make them specific
		     they should take a parameter.
		*/

	    }

	    if (len <= 0) {
		fprintf(stderr, "Invalid operation '%s'\n", opname);
		fclose(fh);
		return -1;
	    }
	    len += magic_offset;
	    if (overlap < len)
		overlap = len;

	} else if (buf[0] != '#' && buf[0] != '\0') {

	    if (sscanf(buf, "extension %63s", r.extension) == 1) {
		/* do nothing */

	    } else if (strncmp(buf, "command ", 8) == 0) {
		r.command = strdup(buf + 8);

	    } else if (strncmp(buf, "postextract ", 12) == 0) {
		r.postextract = strdup(buf + 12);

	    } else if (strncmp(buf, "rename ", 7) == 0) {
		r.rename = strdup(buf + 7);

	    } else if (sscanf(buf, "min_output_file %ld", &r.min_output_file)
		    == 1) {
		/* do nothing */
	    } else if (sscanf(buf, "allow_overlap %d", &r.allow_overlap) == 1) {
		/* do nothing */
	    } else {
		fprintf(stderr, "Invalid line in %s: %s\n", recipefile, buf);
		fclose(fh);
		return -1;
	    }

	}
    }

    if (ferror(fh)) {
	fclose(fh);
	fprintf(stderr, "fgets(): %s: %s\n", recipefile, strerror(errno));
	return -1;
    }

    fclose(fh);

    if (!scanner || !r.command) {
	fprintf(stderr, 
"Invalid recipe file %s.\n"
"It must contain one scanner directive, one command directive and at least\n"
"one match directive\n", recipefile);
	return -1;
    }

    scanner = scanner_add_or_reuse(scanner);
    array_add(&scanner->recipes, &r);

    /* Round overlap to machine word boundary, for faster memcpy */
    overlap +=   sizeof(long) - 1;
    overlap &= ~(sizeof(long) - 1);

    return 0;
}

static void usage(void)
{
    printf(
"Usage: magicrescue [-I FILE] [-M MODE] [-O [+-=][0x]OFFSET] [-b BLOCKSIZE]\n"
"\t-d OUTPUT_DIR -r RECIPE1 [-r RECIPE2 [...]] DEVICE1 [DEVICE2 [...]]\n"
"\n"
"  -b  Only consider files starting at a multiple of BLOCKSIZE.\n"
"  -d  Mandatory.  Output directory for found files.\n"
"  -r  Mandatory.  Recipe name, file or directory.\n"
"  -I  Read input file names from this file (\"-\" for stdin)\n"
"  -M  Produce machine-readable output to stdout.\n"
"  -O  Resume from specified offset (hex or decimal) in the first device.\n"
"\n");
}

int main(int argc, char **argv)
{
    int c;
    struct scanner *global_scanner = NULL;
    char *start_offset = NULL;
    char *file_names_from = NULL;

    array_init(&scanners,    sizeof(struct scanner));
    array_init(&skip_ranges, sizeof(struct range));

    /* Some recipes depend on parsing error messages from various programs */
    putenv("LC_ALL=C");

#ifdef HAVE_GETRLIMIT
    /* Some helper programs will segfault or exhaust memory on malformed
     * input.  This should prevent core files and swap storms */
    {
	struct rlimit rl;
	rl.rlim_cur = rl.rlim_max = 0;
	setrlimit(RLIMIT_CORE, &rl);

	if (getrlimit(RLIMIT_AS, &rl) == 0 && (
		    rl.rlim_max == RLIM_INFINITY ||
		    rl.rlim_max > MAX_MEMORY)) {
	    rl.rlim_cur = rl.rlim_max = MAX_MEMORY;

	    setrlimit(RLIMIT_AS, &rl);
	}
    }
#endif /* HAVE_GETRLIMIT */

    progress.position = 0;
    progress.device[0] = progress.device_basename[0] = '\0';

    buf = malloc(bufsize);

    sprintf(buf, "PATH=%s%s%s:%s", "tools",
#ifdef COMMAND_PATH
	":", COMMAND_PATH,
#else
	"", "",
#endif
	getenv("PATH"));
    putenv(strdup(buf));

    while ((c = getopt(argc, argv, "b:d:r:I:M:O:")) >= 0) {
	switch (c) {
	case 'b': {
	    long boundary = atol(optarg);
	    if (global_scanner)
		scanner_destroy(global_scanner);

	    if (boundary <= 1) {
		global_scanner = NULL;
	    } else {
		if (boundary & (boundary - 1)) {
		    fprintf(stderr,
			    "Error: block size must be a multiple of two\n");
		    return 1;
		}
		global_scanner = scanner_new();
		global_scanner->func = scanner_block;
		global_scanner->param.block = boundary;
	    }
	}
	break; case 'd': {
	    struct stat dirstat;
	    if (stat(optarg, &dirstat) == 0 && dirstat.st_mode & S_IFDIR) {
		output_dir = optarg;
	    } else {
		fprintf(stderr, "Invalid directory %s\n", optarg);
		return 1;
	    }
	}
	break; case 'r': {
	    struct stat st;
	    DIR *dh;
	    if (stat(optarg, &st) == 0 && S_ISDIR(st.st_mode)
		    && (dh = opendir(optarg))) {
		struct dirent *dirent;
		char fullname[PATH_MAX];

		while ((dirent = readdir(dh))) {
		    if (dirent->d_name[0] != '.') {
			snprintf(fullname, PATH_MAX, "%s/%s",
				optarg, dirent->d_name);
			if (parse_recipe(fullname, global_scanner) != 0) {
			    return 1;
			}
		    }
		}
		closedir(dh);

	    } else if (parse_recipe(optarg, global_scanner) != 0) {
		return 1;
	    }
	}
	break; case 'I':
	    file_names_from = optarg;

	break; case 'M':
	    if (strchr(optarg, 'i'))
		machine_output |= OUT_I;
	    if (strchr(optarg, 'o'))
		machine_output |= OUT_O;
	    setvbuf(stdout, NULL, _IOLBF, 0);

	break; case 'O':
	    start_offset = optarg;

	break; default:
	    fprintf(stderr, "Error parsing options.\n");
	    usage();
	    return 1;
	}
    }

    if (array_count(&scanners) == 0 || !output_dir) {
	usage();
	return 1;
    }

    name_mode = (argc-optind>5 || file_names_from) ? MODE_FILES : MODE_DEVICE;

    signal(SIGINT,  signal_beforedeath);
    signal(SIGTERM, signal_beforedeath);
    signal(SIGSEGV, signal_beforedeath);
    signal(SIGPIPE, signal_beforedeath);
    
    while (optind < argc) {
	scan_disk(argv[optind++], start_offset);
	start_offset = NULL;
    }

    if (file_names_from) {
	char inputfile[PATH_MAX];
	FILE *fh = strcmp(file_names_from, "-") == 0 ? stdin : 
		fopen(file_names_from, "r");
	
	if (!fh) {
	    fprintf(stderr, "Opening %s: %s\n",
		    file_names_from, strerror(errno));
	    return 1;
	}

	while (fgets(inputfile, sizeof inputfile, fh)) {
	    inputfile[strlen(inputfile)-1] = '\0'; /* kill newline */
	    scan_disk(inputfile, NULL);
	}
	fclose(fh);
    }

    scanners_free();
    array_destroy(&skip_ranges);
    free(buf);
    return 0;
}

/* vim: ts=8 sw=4 noet tw=80
 */
